src/utils/net.rs

```diff
@@ -40,9 +52,69 @@ fn enable_reuse(sock: &Socket) -> io::Result<()> {
     Ok(())
 }
 
+/// Socket that can accept and send data from either a local ipv4 address or ipv6 address.
+pub struct DualStackLocalSocket {
+    v4: UdpSocket,
+    v6: UdpSocket,
+}
+
+impl DualStackLocalSocket {
+    pub fn new(port: u16) -> Result<DualStackLocalSocket> {
+        // if ephemeral port, make sure they are on the same ports.
+        if port == 0 {
+            let v4 = socket_with_reuse_and_address((std::net::Ipv4Addr::UNSPECIFIED, port).into())?;
+            let port = v4.local_addr()?.port();
+
+            return Ok(Self {
+                v4,
+                v6: socket_with_reuse_and_address((std::net::Ipv6Addr::UNSPECIFIED, port).into())?,
+            });
+        }
+
+        Ok(Self {
+            v4: socket_with_reuse_and_address((std::net::Ipv4Addr::UNSPECIFIED, port).into())?,
+            v6: socket_with_reuse_and_address((std::net::Ipv6Addr::UNSPECIFIED, port).into())?,
+        })
+    }
+
+    pub async fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
+        let mut v4_buf = vec![0; buf.len()];
```

Comment:
Is there a better way than creating a whole new buffer each time? I can't reuse the same buffer in both places as Rust won't let you borrow something mutable more than once.

Reply:
I think a reasonable first implementation would be to have two member buffers, and have the API return the slice (returning a slice should work here, but you might need to copy into a vec). This means slightly more initial memory, but will reduce the number of allocations.

Reply:
I like it. If I need to copy to a vec, I might also look at returning an enum on which buffer was populated, so we don't have to create a whole vec each time.